File: Data_Structures/Queue/PriorityQueue/JavaScript/PriorityQueue.js
/*
 * PriorityQueue - Min-heap based priority queue
 *
 * Operations: push O(log n), pop O(log n), peek O(1)
 * Smallest value has highest priority.
 */

class PriorityQueue {
    constructor() {
        this._heap = [];
    }

    isEmpty() { return this._heap.length === 0; }
    size()    { return this._heap.length; }

    _swap(i, j) {
        [this._heap[i], this._heap[j]] = [this._heap[j], this._heap[i]];
    }

    _heapifyUp(i) {
        while (i > 0) {
            const parent = Math.floor((i - 1) / 2);
            if (this._heap[i] < this._heap[parent]) {
                this._swap(i, parent);
                i = parent;
            } else {
                break;
            }
        }
    }

    _heapifyDown(i) {
        const n = this._heap.length;
        let smallest = i;
        const left  = 2 * i + 1;
        const right = 2 * i + 2;
        if (left  < n && this._heap[left]  < this._heap[smallest]) smallest = left;
        if (right < n && this._heap[right] < this._heap[smallest]) smallest = right;
        if (smallest !== i) {
            this._swap(i, smallest);
            this._heapifyDown(smallest);
        }
    }

    push(value) {
        this._heap.push(value);
        this._heapifyUp(this._heap.length - 1);
    }

    pop() {
        if (this.isEmpty()) throw new Error("Priority queue is empty");
        if (this.size() === 1) return this._heap.pop();
        const min = this._heap[0];
        this._heap[0] = this._heap.pop();
        this._heapifyDown(0);
        return min;
    }

    peek() {
        if (this.isEmpty()) throw new Error("Priority queue is empty");
        return this._heap[0];
    }
}

const pq = new PriorityQueue();
pq.push(30);
pq.push(10);
pq.push(50);
pq.push(20);

console.log("Peek (min):", pq.peek());
while (!pq.isEmpty()) {
    console.log("Pop:", pq.pop());
}

File: Data_Structures/Queue/PriorityQueue/C++/PriorityQueue.cc
/*
 * PriorityQueue - Min-heap based priority queue
 *
 * Operations: push O(log n), pop O(log n), peek O(1)
 * Smallest value has highest priority.
 */

#include <bits/stdc++.h>
using namespace std;

class PriorityQueue {
private:
    vector<int> heap;

    void heapifyUp(int i) {
        while (i > 0) {
            int parent = (i - 1) / 2;
            if (heap[i] < heap[parent]) {
                swap(heap[i], heap[parent]);
                i = parent;
            } else {
                break;
            }
        }
    }

    void heapifyDown(int i) {
        int n = heap.size();
        int smallest = i;
        int left  = 2 * i + 1;
        int right = 2 * i + 2;
        if (left  < n && heap[left]  < heap[smallest]) smallest = left;
        if (right < n && heap[right] < heap[smallest]) smallest = right;
        if (smallest != i) {
            swap(heap[i], heap[smallest]);
            heapifyDown(smallest);
        }
    }

public:
    bool isEmpty() const { return heap.empty(); }
    int  size()    const { return (int)heap.size(); }

    void push(int value) {
        heap.push_back(value);
        heapifyUp((int)heap.size() - 1);
    }

    int pop() {
        if (isEmpty()) throw out_of_range("Priority queue is empty");
        int min = heap[0];
        heap[0] = heap.back();
        heap.pop_back();
        if (!isEmpty()) heapifyDown(0);
        return min;
    }

    int peek() const {
        if (isEmpty()) throw out_of_range("Priority queue is empty");
        return heap[0];
    }
};

int main() {
    PriorityQueue pq;

    pq.push(30);
    pq.push(10);
    pq.push(50);
    pq.push(20);

    cout << "Peek (min): " << pq.peek() << endl;
    while (!pq.isEmpty()) {
        cout << "Pop: " << pq.pop() << endl;
    }

    return 0;
}
